src/main/java/chatbot/Duke.java

```diff
@@ -0,0 +1,225 @@
+ */
+class EmptyDeadlineException extends Exception {
+
+  public EmptyDeadlineException() {
+    super("Deadline cannot have empty description.");
+  }
+}
+
+/**
+ * Thrown to indicate that the description for an Event task cannot be empty.
+ */
+class EmptyEventException extends Exception {
+
+  public EmptyEventException() {
+    super("Event cannot have empty description.");
+  }
+}
+
+/**
+ * Thrown to indicate that the user input is invalid.
+ */
+class InvalidException extends Exception {
+
+  public InvalidException() {
+    super("Sorry, invalid input.");
+  }
+}
+
+/**
+ * Represents a generic task with a description and a completion status.
+ */
+class Task {
+
+  private String description;
+  private boolean isDone;
+
+  public Task(String description) {
+    assert description != null &&
+    !description.trim().isEmpty() : "Task description cannot be null or empty";
+    this.description = description;
+    this.isDone = false;
+  }
+
+  public String getStatusIcon() {
+    return (isDone ? "[X] " : "[ ] "); // mark done task with X
+  }
+
+  public void markAsDone() {
+    this.isDone = true;
+  }
+
+  public void markAsUndone() {
+    this.isDone = false;
+  }
+
+  @Override
+  public String toString() {
+    return (this.getStatusIcon() + this.description);
+  }
+
+  public String getDescription() {
+    return description;
+  }
+}
+
+/**
+ * Represents a Todo task.
+ */
+class Todo extends Task {
+
+  public Todo(String description) {
+    super(description);
+  }
+
+  @Override
+  public String toString() {
+    return ("[T]" + super.toString());
+  }
+}
+
+/**
+ * Represents a Deadline task with a due date.
+ */
+class Deadline extends Task {
+
+  protected String by;
+
+  public Deadline(String description, String dateString) {
+    super(description);
+    assert dateString != null : "Deadline date cannot be null";
+    LocalDate date = LocalDate.parse(dateString);
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+    this.by = date.format(formatter);
+  }
+
+  @Override
+  public String toString() {
+    return "[D]" + super.toString() + " (by: " + by + ")";
+  }
+}
+
+/**
+ * Represents an Event task with a start and end time.
+ */
+class Event extends Task {
+
+  protected String from;
+  protected String to;
+
+  public Event(String description, String from, String to) {
+    super(description);
+    assert from != null && to != null : "Event times cannot be null";
+    this.from = from;
+    this.to = to;
+  }
+
+  @Override
+  public String toString() {
+    return (
+      "[E]" + super.toString() + " (from: " + from + ") " + "to: " + to + ")"
+    );
+  }
+}
+
+/**
+ * The main class for the chatbot application. It initializes and runs the chatbot,
+ * handling the primary control flow of the application.
+ */
+public class Duke {
+
+  private TaskList myList;
+  private Storage st;
+  private Ui ui;
+  private Parser p;
```

Comment:
Please use full nouns for the variable name. Writing parser instead of p is fine!